### PR TITLE
[Datasets] Add fast file metadata provider and refactor Parquet datasource

### LINKED ISSUE #22910: [Feature] Ray dataset loading large list of parquet files is extremely slow
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Trying to use ray to run inference on our big dataset. And we are using ray dataset to load the dataset. The read operation is quite slow on 2k files, and increases linearly proportional to number of files, as our goal is to process 100k files. 

Here is a simple code snippet to "reproduce" the "problem".
```
import gcsfs
import ray

fs = gcsfs.GCSFileSystem()
len(file_names)
# 2048
%time ds = ray.data.read_parquet(file_names, filesystem=fs, parallelism=len(file_names))
# 4 minute
```

### Use case

After some poking around, the bottleneck appears to be using [arrow.parquet.ParquetDataset](https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/parquet_datasource.py#L92) as the basis for ray.dataset. It's a single machine, sequential operation. And have no much room to user to scale up. 

```
import pyarrow.parquet as pq
len(file_names)
# 2048
%time ds = pq.ParquetDataset(file_names, filesystem=fs)
# Wall time: 2min 55s
```
appears majority of the time is spend on this operation.  There are some room for improving pyarrow by using multiple cpu, but it doesn't help in our case :(  
```
import pyarrow as pa
pa.set_cpu_count(10)
%time ds = pq.ParquetDataset(file_names, filesystem=fs)
# Wall time: 3min 56s
```

We've used apache beam before for loading dataset for inference. Which doesn't have this type of problem, since it's using [arrow.parquet.ParquetFile](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/parquetio.py#L332), and handles the filenames using beam distributed data abstraction. This is a much more efficient/smart approach, as it would process the files in parallel hence adding machine would definitely help.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by pdames**
## Why are these changes needed?

Adds a fast file metadata provider that trades comprehensive file metadata collection for speed of metadata collection, and which also disabled directory path expansion which can be very slow on some cloud storage service providers. This PR also refactors the Parquet datasource to be able to take advantage of both these changes and the content-type agnostic partitioning support from https://github.com/ray-project/ray/pull/23624. 

This is the second PR of a series originally proposed in https://github.com/ray-project/ray/pull/23179.

## Related issue number

Partially resolves https://github.com/ray-project/ray/issues/22910.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/file_based_datasource.py [hunk 1]
```diff
@@ -237,6 +244,32 @@ def expand_paths(
         return expanded_paths, file_sizes
 
 
+class FastFileMetadataProvider(DefaultFileMetadataProvider):
```
Comment (jianoaix) [inline python/ray/data/datasource/file_based_datasource.py:247]: I find it weird to have those in a *_datasource.py file. Would it make more sense for all those subclasses of FileMetadataProvider to be placed in filed_meta_provider.py? If not, why?

File: python/ray/data/datasource/file_based_datasource.py [hunk 2]
```diff
@@ -237,6 +244,32 @@ def expand_paths(
...
+    path expansion and file size collection. While this performance improvement may be
+    negligible for local filesystems, it can be substantial for cloud storage service
+    providers.
+
+    This should only be used when all input paths are known to be files.
+    """
+
+    def expand_paths(
+        self,
+        paths: List[str],
```
Comment (jianoaix) [inline python/ray/data/datasource/file_based_datasource.py:260]: Do all paths need to be for the same block?
If not, why does it make sense to return BlockMetadata as value?

File: python/ray/data/datasource/file_based_datasource.py (hunk 1 above)
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py:247]: I had considered the same, and I think it would be a good idea to move them all over to `file_metadata_provider.py` as part of this PR.

From the perspective of an end-user, they should be able to use `from ray.data.datasource import FastFileMetadataProvider` regardless of which file we put the class inside of, so this shouldn't change much from their perspective.

From the perspective of a Ray Data maintainer, the current organization largely grew out of organic growth of the code, where one-off utility functions have gradually evolved into more generic/extensible classes over time, and thus follow existing conventions of grouping utility class impls with the datasource that uses them. Thus, all file metadata providers given in `file_based_datasource.py` are meant to be used with file-based datasources, while all file metadata providers in `parquet_datasource.py` are meant to be used specifically with the `ParquetDatasource`.

The primary downside of continuing to preserve this type of grouping is that core dependencies like `file_based_datasource.py` starts to grow too large over time (sitting now at >700 LoC), which makes it harder to grok the important parts at a glance or to understand during a quick top-to-bottom read. 

If we were to move all file metadata provider implementations into `file_metadata_provider.py`, I'd expect it to be just over 300 LoC with all classes organized around the central purpose of providing file metadata. This better normalizes the size of our "large files" and keeps them more focused on a single purpose, so I'm slightly in favor of making this move now if we're in agreement that it's beneficial.

We probably also should do the same with block write path providers in a follow-up PR.

File: python/ray/data/datasource/file_based_datasource.py (hunk 2 above)
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py:260]: I assume this comment is meant to apply to the core `__call__` and `_get_block_metadata` methods, which do require that all paths provided are part of the same block. I think this could be made more clear with some doc string updates.

File: python/ray/data/datasource/file_based_datasource.py (hunk 1 above)
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py]: I refactored this in the latest commit - let me know what you think.

File: python/ray/data/datasource/file_based_datasource.py (hunk 2 above)
Comment (pdames) [reply python/ray/data/datasource/file_based_datasource.py]: I've added some clarification to the docstrings here in the latest commit.

File: python/ray/data/datasource/parquet_base_datasource.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+from ray.data.block import BlockAccessor
+from ray.data.datasource.file_based_datasource import (
+    FileBasedDatasource,
+    _resolve_kwargs,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ParquetBaseDatasource(FileBasedDatasource):
```
Comment (jianoaix) [inline python/ray/data/datasource/parquet_base_datasource.py:21]: This class has only internal methods and also has only one subclass -- can it be folded into its subclass ParquetDatasource? My concern is it's a bit overskill for this case.

File: python/ray/data/datasource/file_meta_provider.py [hunk 4]
```diff
@@ -46,3 +53,271 @@ def __call__(
...
+            f"directories or if file size collection is required, try rerunning this "
+            f"read with `meta_provider=DefaultFileMetadataProvider()`."
+        )
+        import numpy as np
+
+        return paths, np.empty(len(paths), dtype=object)
+
+
+@DeveloperAPI
+class ParquetMetadataProvider(FileMetadataProvider):
```
Comment (jianoaix) [inline python/ray/data/datasource/file_meta_provider.py:210]: I have similar concern of whether this layer is needed, i.e. can DefaultParquetMetadataProvider directly subclass FileMetadataProvider?

File: python/ray/data/datasource/file_meta_provider.py [hunk 5]
```diff
@@ -46,3 +53,271 @@ def __call__(
         **kwargs,
     ) -> BlockMetadata:
         return self._get_block_metadata(paths, schema, **kwargs)
+
+
+@DeveloperAPI
+class BaseFileMetadataProvider(FileMetadataProvider):
```
Comment (jianoaix) [inline python/ray/data/datasource/file_meta_provider.py:59]: Similarly for this class?

**Review by jianoaix — COMMENTED** [review]
Looking good overall. My question is whether we can make the chain of inheritance shorter/simpler?

File: python/ray/data/datasource/file_based_datasource.py (hunk 1 above)
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py]: Thanks, I think it makes sense for subclasses of FileMetadataProvider live in file_metadata_provider.py.

File: python/ray/data/datasource/file_based_datasource.py (hunk 2 above)
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py]: Thanks.
@clarkzinzow Is it true that we have multiple files to load into a single block?

File: python/ray/data/datasource/parquet_base_datasource.py (hunk 3 above)
Comment (pdames) [reply python/ray/data/datasource/parquet_base_datasource.py:21]: I think that depends on what folding it into `ParquetDatasource` actually means but, if we mean getting rid of this class altogether, I don't think that will work moving forward. Some degree of refactoring may be possible, but we will need to separate this class in either this PR or the next PR since the upcoming Parquet bulk file reader API needs to use `ParquetBaseDatasource` directly. This API will not be able to use `ParquetDatasource` directly due to its `prepare_read` method override being the root cause of subsequent scalability/performance issues when consuming a large number of Parquet files cited in https://github.com/ray-project/ray/issues/22910. 

So we can delay separation of this class until the next PR, but I think it still needs to happen.

File: python/ray/data/datasource/file_meta_provider.py (hunk 4 above)
Comment (pdames) [reply python/ray/data/datasource/file_meta_provider.py:210]: I have a slight preference to preserve the current layout since the intent is for anyone providing a custom `ParquetDatasource` metadata provider to implement the interface signature established in `ParquetMetadataProvider` and, in particular:
1. Only implement metadata prefetching if their use-case requires it. 
2. Carefully consider the best way to fetch block metadata for their use-case rather than just keep a default implementation.

For example, my `CachedFileMetadataProvider` at https://github.com/pdames/deltacat/blob/edec6159c5acda3ede15653b4e92aaa45a43206f/deltacat/io/aws/redshift/redshift_datasource.py#L67-L70 inherits from `ParquetMetadataProvider`, does not require metadata prefetching, and uses a different implementation for getting block metadata from a prebuilt cache. I also expect this to be the general case for most upcoming data warehouse and data catalog integrations with Ray Datasets beyond just the Amazon Redshift integration that uses it here.

So, in summary, my slight preference is to keep `DefaultParquetMetadataProvider` set aside as an internal implementation detail exposed to Ray Data maintainers (and to continue to keep the `@DeveloperAPI` label excluded from this class), and for `ParquetMetadataProvider` to be exposed to end-users creating their own `ParquetDatasource` metadata providers.

File: python/ray/data/datasource/file_meta_provider.py (hunk 5 above)
Comment (pdames) [reply python/ray/data/datasource/file_meta_provider.py:59]: I have the same slight preference to preserve the structure here for this class as above - namely that `BaseFileMetadataProvider` remain the developer API that end-users implement their own custom `FileBasedDatasource` metadata providers against, that `DefaultFileMetadataProvider` and `FastFileMetadataProvider` remain more hidden Ray Data implementation details, and that implementors think carefully about if they need path expansion (plus how they would best implement it if so, since a metadata provider specific to a given cloud service or data platform could be much more performant/scalable here) and how they should best retrieve block metadata for their use-case.

File: python/ray/data/datasource/parquet_base_datasource.py (hunk 3 above)
Comment (jianoaix) [reply python/ray/data/datasource/parquet_base_datasource.py:21]: Ok, thanks for explanation. Let's keep it then.

File: python/ray/data/datasource/file_meta_provider.py (hunk 5 above)
Comment (pdames) [reply python/ray/data/datasource/file_meta_provider.py:59]: In both cases, the other part of this pertains to breaking backwards compatibility, since any custom metadata providers that may already be inheriting from the existing `BaseFileMetadataProvider` and `ParquetMetadataProvider` classes would be broken by this change.

File: python/ray/data/datasource/file_meta_provider.py (hunk 4 above)
Comment (jianoaix) [reply python/ray/data/datasource/file_meta_provider.py:210]: SG, thanks.

**Review by jianoaix — APPROVED** [review]
Thank you Patrick, LG!

**Review by clarkzinzow — APPROVED** [review]
LGTM, awesome work!

**Comment by clarkzinzow** [comment]
Hey @pdames, just reverted a [commit](https://github.com/ray-project/ray/commit/15d66a8dd71c98b1224cfbe0d579537e692841e1) that was breaking Datasets CI job, could you rebase on master one more time? 🙏

**Comment by pdames** [comment]
> Hey @pdames, just reverted a [commit](https://github.com/ray-project/ray/commit/15d66a8dd71c98b1224cfbe0d579537e692841e1) that was breaking Datasets CI job, could you rebase on master one more time? 🙏

Done!

**Comment by pdames** [comment]
@clarkzinzow Looks like the remaining CI failures are unrelated. Could you give this a final pass through?

**Comment by clarkzinzow** [comment]
LGTM, merging!
